[prototype] sql: enable secondary tenants to write zone configurations

arulajmani opened the pull request:
Fixes #67014.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67323)
<!-- Reviewable:end -->

irfansharif commented on pkg/ccl/logictestccl/testdata/logic_test/tenant_unsupported:
```diff
@@ -53,9 +53,6 @@ SELECT * FROM crdb_internal.zones
 statement error operation is unsupported in multi-tenancy mode
 SHOW ZONE CONFIGURATION FOR TABLE kv
 
-statement error operation is unsupported in multi-tenancy mode
```
Move this statement test to `pkg/ccl/logictestccl/testdata/logic_test/tenant` instead. I'd delete the `zone_config_tenants` test file for now. Also -- why does it show up as a binary file on github/locally? Try `$ file <path to zone_config_tenants>`. Is it because we're printing out the proto bytes directly? Seems bad for a logic test, printing out number of rows would be better. Better yet, test what's visible through `SHOW ZONE CONFIGURATIONS` instead. 

I think the only behavioural difference between secondary/host tenants will be around not being able to set/view the distinguished zone configs (liveness, meta -- see RFC), for those it makes sense to use a separate tenant-specific file. Could do it in this PR, or add a TODO(zcfgs-pod) somewhere.

```
statement ok
ALTER RANGE DEFAULT CONFIGURE ZONE USING num_replicas = 5;
```

irfansharif commented on pkg/clusterversion/cockroach_versions.go:
```diff
@@ -304,6 +304,11 @@ const (
 
+	// ZonesTableForSecondaryTenants adds the zones system table for all secondary
+	// tenants. This doesn't affect the system tenant, as the system tenant
+	// already has the zones table.
```
Just the first sentence is sufficient, given the version name. Also s/the zones system table/system.zones.

irfansharif commented on pkg/config/keys.go:
```diff
@@ -13,6 +13,7 @@ package config
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 )
 
 // MakeZoneKeyPrefix returns the key prefix for id's row in the system.zones
```
This _function_ (as opposed to a method on the `SystemConfig` type) should thread in a codec.

irfansharif commented on pkg/config/keys.go:
```diff
@@ -21,9 +22,15 @@ func MakeZoneKeyPrefix(id SystemTenantObjectID) roachpb.Key {
+	return MakeZoneKey(keys.SystemSQLCodec, descpb.ID(id))
 }
 
 // DecodeSystemTenantObjectID decodes the object ID for the system-tenant from
```
This function should also thread in a codec. Does it have to return this SystemTenantObjectID type?

irfansharif replied:
Better yet, can we get rid of the SystemTenantObjectID type entirely? Instead we could ensure we're using a codec + descpb.ID everywhere, specifying the SystemSQLCodec explicitly whenever needed.

irfansharif commented on pkg/config/keys.go:
```diff
@@ -21,9 +22,15 @@ func MakeZoneKeyPrefix(id SystemTenantObjectID) roachpb.Key {
+
+// MakeZoneKeyForSystemTenant returns the key for the given system tenant ID's
+// entry in the system.zones table.
+func MakeZoneKeyForSystemTenant(id SystemTenantObjectID) roachpb.Key {
```
Let's get rid of this `config.MakeZoneKeyForSystemTenant` API, `MakeZoneKey` is sufficient. For callers that want to only address the SystemTenant, lets have them thread in `SystemSQLCodec`.

irfansharif commented on pkg/config/system.go:
```diff
@@ -537,7 +537,7 @@ func (s *SystemConfig) systemTenantTableBoundarySplitKey(
 				return nil
 			}
```
Elsewhere in this file, DecodeKeyIntoZoneIDAndSuffix should thread in a codec.

irfansharif commented on pkg/migration/migrations/zones.go:
```diff
@@ -0,0 +1,35 @@
+	"github.com/cockroachdb/cockroach/pkg/sqlmigrations"
+)
+
+// zonesTableForSecondaryTenants adds system.zones to existing secondary
```
s/existing//

irfansharif commented on pkg/migration/migrations/zones.go:
```diff
@@ -0,0 +1,35 @@
+func zonesTableForSecondaryTenants(
+	ctx context.Context, _ clusterversion.ClusterVersion, d migration.TenantDeps,
+) error {
+	// We don't need to add system.zones for the host tenant as it should already
```
[nit] Move the comment within the if condition.

irfansharif commented on pkg/migration/migrations/migrations.go:
```diff
@@ -71,6 +71,11 @@ var migrations = []migration.Migration{
 		spanConfigurationsTokensTableMigration,
 	),
+	migration.NewTenantMigration(
+		"add system.zones table to secondary tenants",
```
s/to/for

irfansharif commented on pkg/sql/crdb_internal.go:
```diff
@@ -3229,7 +3228,14 @@ CREATE TABLE crdb_internal.zones (
 			fullZone := configProto
-			if err := completeZoneConfig(&fullZone, config.SystemTenantObjectID(tree.MustBeDInt(r[0])), getKey); err != nil {
+			// TODO(zcfg-pod): We currently lack introspection for zone configurations
+			// for secondary tenants.
```
Why? We have the codec available to us: `p.ExecCfg().Codec`.

irfansharif replied:
Shouldn't it "just work" if we got rid of the `if !p.ExecCfg().Codec.ForSystemTenant()` check above?

irfansharif commented on pkg/sql/drop_index.go:
```diff
@@ -282,7 +281,9 @@ func (p *planner) dropIndexByName(
 	// necessary for the system tenant, because secondary tenants do not have
 	// zone configs for individual objects.
 	if p.ExecCfg().Codec.ForSystemTenant() {
```
Don't we want to get rid of this check? (+ remove the comment above)

irfansharif commented on pkg/sql/set_zone_config.go:
```diff
@@ -185,9 +184,6 @@ func (p *planner) SetZoneConfig(ctx context.Context, n *tree.SetZoneConfig) (pla
 	if err := checkPrivilegeForSetZoneConfig(ctx, p, n.ZoneSpecifier); err != nil {
 		return nil, err
 	}
-	if !p.ExecCfg().Codec.ForSystemTenant() {
```
We need to remove these `if [..]Codec.ForSystemTenant()` checks below, for `RemoveIndexZoneConfigs`. I think we'd benefit from a general audit of all these cases. Maybe lets just try and port over the system tenant zone config logic tests to the tenants right now, almost all of them should apply (save for the distinguished ranges, and for replication reports). That should shake out all these tests.

irfansharif commented on pkg/sql/show_zone_config.go:
```diff
@@ -122,7 +122,7 @@ func getShowZoneConfigRow(
 		}
 	}
 
```
If we remove the `if !p.ExecCfg().Codec.ForSystemTenant() {` check in `ShowZoneConfig`, we should "just" get introspection.

irfansharif commented on pkg/sql/zone_config.go:
```diff
@@ -55,22 +55,17 @@ var errNoZoneConfigApplies = errors.New("no zone config applies")
 // already is a parent and we'll not decode a descriptor.
 func getZoneConfig(
-	id config.SystemTenantObjectID,
+	id descpb.ID,
```
[nit] Swap the order of arguments, codec typically precedes what's being decoded through it.

irfansharif commented on pkg/sql/zone_config.go:
```diff
@@ -129,6 +126,14 @@ func getZoneConfig(
 
+	// TODO(arul): For now, secondary tenants get this static default zone config.
+	//  We probably want to seed system.zones with a default zone configuration
+	//  eventually though.
```
Instead of this, let's just seed it now. Wouldn't it just be around `p.CreateTenant` in `pkg/sql/tenant.go`?

irfansharif commented on pkg/sql/zone_config.go:
```diff
@@ -189,17 +194,30 @@ func zoneConfigHook(
 	cfg *config.SystemConfig, id config.SystemTenantObjectID,
 ) (*zonepb.ZoneConfig, *zonepb.ZoneConfig, bool, error) {
 	getKey := func(key roachpb.Key) (*roachpb.Value, error) {
+		// TODO(arul): clean this up
```
Huh, why did we need this? 

I think generally keeping around this `SystemTenantObjectID` type, compared to other instances where we're passing in the codec + desc ID, and me not fully being able to appreciate the future of this `SystemConfig` type becoming makes some of these invariants non-obvious to me. 

This is pretty hairy code, and trying to untangle it to make the system tenant less distinguished is difficult, but I'd benefit from a high-level direction you see these type taking. Is that apparent to you? No worries if not, just wanted to share the confusion. A lot of the "host tenant only" stuff is in the kvserver/reports/reporter, which will also later end up generalizing to tenants.

irfansharif commented on pkg/sql/set_zone_config.go:
```diff
@@ -673,18 +680,21 @@ func (n *setZoneConfigNode) startExec(params runParams) error {
-			if err != nil {
-				return err
-			}
+			// TODO(arul): Get rid of this once tenants have access to node localites.
```
```
// TODO(zcfgs-pod): We want to validate for secondary tenants as
// well, but they don't (yet) have access to node localites.
```

irfansharif reviewed (commented):
Spot checking a few more places where we missed the `[...]Codec.ForSystemTenant` checks:

https://github.com/cockroachdb/cockroach/blob/36ed61e139271756805545ae5939c40fb32e7e45/pkg/sql/schema_changer.go#L1382-L1385

https://github.com/cockroachdb/cockroach/blob/36ed61e139271756805545ae5939c40fb32e7e45/pkg/sql/schema_changer.go#L2171

https://github.com/cockroachdb/cockroach/blob/36ed61e139271756805545ae5939c40fb32e7e45/pkg/sql/gcjob/descriptor_utils.go#L65-L68

---

This also needs to be updated. 

https://github.com/cockroachdb/cockroach/blob/36ed61e139271756805545ae5939c40fb32e7e45/pkg/sql/catalog/bootstrap/metadata.go#L359-L363

For the zone configs below, we'll want to only add for the system tenant (see RFC):

https://github.com/cockroachdb/cockroach/blob/36ed61e139271756805545ae5939c40fb32e7e45/pkg/sql/catalog/bootstrap/metadata.go#L372-L374

---

See comments below, I think we're close enough to introspect written zone configs that we should just do it in this PR. To get full coverage for all these checks above, we could also start running existing zone config logic tests against tenants and see what breaks. Some parts will need to get segregated of course (like all replication reports, or zone configs for liveness/meta ranges, etc.), but most of it will stay the same. What do you think? I'd personally prefer happy iterating on this PR to iron out most of these SQL-side kinks, but if you prefer to follow up in separate PRs, that's fine too. Though I do think we should get better coverage of these checks in this PR if possible.

arulajmani reviewed (commented):
> To get full coverage for all these checks above, we could also start running existing zone config logic tests against tenants and see what breaks. Some parts will need to get segregated of course (like all replication reports, or zone configs for liveness/meta ranges, etc.), but most of it will stay the same. What do you think? I'd personally prefer happy iterating on this PR to iron out most of these SQL-side kinks, but if you prefer to follow up in separate PRs, that's fine too. Though I do think we should get better coverage of these checks in this PR if possible.

Yeah, I have this whack-a-mole going on on a separate branch for now, but I'm happy to tack on the commits to this PR itself. I'll ping on this PR again once I've found a decent stopping point. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-:-Me6pr1ODNtkyI8UFsiA:bex1blv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @irfansharif)

___
*[pkg/config/keys.go, line 36 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSq-McpWkamiJMV-r3-36:-Me735Wc3tE5an3MOWGk:bbdtfpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/config/keys.go#L36)):*
> Better yet, can we get rid of the SystemTenantObjectID type entirely?

See my comment below about this. I don't think we can as part of this PR, but we should work towards this eventually. 

> This function should also thread in a codec. Does it have to return this SystemTenantObjectID type?

Considering this thing is only called through from codepaths in kvserver, I was thinking generalizing that stuff would serve as a natural place to do this. Do you think that's reasonable? 
___
*[pkg/sql/zone_config.go, line 131 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSrRQKA-mVg5nqt-r3-131:-Me72UUqEaigY-b4GqTR:bu0usw) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/sql/zone_config.go#L131)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Instead of this, let's just seed it now. Wouldn't it just be around `p.CreateTenant` in `pkg/sql/tenant.go`?
</blockquote></details>

Thanks for the pointer, the only reason I hadn't got it in here was because I wasn't sure how these things are initialized for tenants. 
___
*[pkg/sql/zone_config.go, line 197 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSrRQKA-mVg5nqt-r3-197:-Me6rDt47-HL08Ky-T8P:b-pi565z) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/sql/zone_config.go#L197)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Huh, why did we need this? 

I think generally keeping around this `SystemTenantObjectID` type, compared to other instances where we're passing in the codec + desc ID, and me not fully being able to appreciate the future of this `SystemConfig` type becoming makes some of these invariants non-obvious to me. 

This is pretty hairy code, and trying to untangle it to make the system tenant less distinguished is difficult, but I'd benefit from a high-level direction you see these type taking. Is that apparent to you? No worries if not, just wanted to share the confusion. A lot of the "host tenant only" stuff is in the kvserver/reports/reporter, which will also later end up generalizing to tenants. 
</blockquote></details>

I should've done a better job adding words to that TODO haha. I'll try to add more context on what the state was before my change, why I chose this as the stopping point, and where I see this going once we have this whole thing wired up. 

At a high level, all zone config code is organized in a way that it goes through this `getZoneConfig` function. When KV accesses it using this hook, `getKey` reads the system config span. Other SQL code does transactional reads. Previously, SQL code was gated to only let the system tenant through. As such, we could cast descriptor IDs to `SystemTenantObjectIDs`. This is no longer true after my change, so I've generalized all code in SQL to require codecs and work for arbitrary object IDs. 

This function here, that KV calls, is one of the few places in SQL that still uses `SystemTenantObjectID`. Once we've generalized things in kvserver and it no longer calls into SQL (and instead reference the span configs table), we'll get rid of this whole `zoneConfigHook` entirely. Until then, I have this assertion baked in here. 

Once we've generalized everything in kvserver/reporter I think we can get rid of `SystemTenantObjectID` in its entirety. That's why removing (most) occurrences in the sql package felt like a good stopping point. 

Separately, one of the things that makes this zone config code super hairy is this `getKey` business and everything funnelling to `getZoneConfig`. I'd really like to simplify all this once we've gotten rid of this `zoneConfigHook`, but that'll be mostly cleanup.  


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
RFAL 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-:-MeQnXjc3T3iSBVDcVK5:b-828fno)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @irfansharif)

___
*[pkg/clusterversion/cockroach_versions.go, line 309 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSpUEp9isMC9erx-r3-309:-MeQKqTa-f5ro48PTTBD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/clusterversion/cockroach_versions.go#L309)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Just the first sentence is sufficient, given the version name. Also s/the zones system table/system.zones.
</blockquote></details>

Done.
___
*[pkg/config/keys.go, line 32 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSq-McpWkamiJMV-r3-32:-MeQh2DBB0aiY84Qxqfm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/config/keys.go#L32)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Let's get rid of this `config.MakeZoneKeyForSystemTenant` API, `MakeZoneKey` is sufficient. For callers that want to only address the SystemTenant, lets have them thread in `SystemSQLCodec`.
</blockquote></details>

Done.
___
*[pkg/config/system.go, line 538 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSq-McpWkamiJMW-r3-538:-MeQeVSM9RL_U7OZk8tr:b-w1sbee) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/config/system.go#L538)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Elsewhere in this file, DecodeKeyIntoZoneIDAndSuffix should thread in a codec.
</blockquote></details>

I was hoping to isolate codec threading to just the SQL code required to set/introspect zone configs. Considering `DecodeKeyIntZoneIDAndSuffix` is primarily called from the replication report code, I think it makes sense to thread in the codec once we're tackling that. Wdyt? 
___
*[pkg/migration/migrations/migrations.go, line 75 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSpUEp9isMC9erz-r3-75:-MeQMImL2EDCGlFw9RsC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/migration/migrations/migrations.go#L75)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/to/for
</blockquote></details>

Done.
___
*[pkg/migration/migrations/zones.go, line 22 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSpUEp9isMC9es--r3-22:-MeQMSPR0xq96bfCZS5_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/migration/migrations/zones.go#L22)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/existing//
</blockquote></details>

Done.
___
*[pkg/migration/migrations/zones.go, line 27 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSpUEp9isMC9es--r3-27:-MeQMWIP6FQAkFNjzdYT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/migration/migrations/zones.go#L27)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Move the comment within the if condition.
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 3232 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSq-McpWkamiJMb-r3-3232:-MeQhWpi9K0sxQpsFt92:bnekrja) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/sql/crdb_internal.go#L3232)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Shouldn't it "just work" if we got rid of the `if !p.ExecCfg().Codec.ForSystemTenant()` check above?
</blockquote></details>

Tacked on in a separate commit. 
___
*[pkg/sql/drop_index.go, line 283 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSrRQKA-mVg5nqp-r3-283:-MeQNMuD-hXlgGIySvNK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/sql/drop_index.go#L283)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Don't we want to get rid of this check? (+ remove the comment above)
</blockquote></details>

Done.
___
*[pkg/sql/set_zone_config.go, line 683 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSrRQKA-mVg5nqr-r3-683:-MeQOddm74naxTtuCN4u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/sql/set_zone_config.go#L683)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

```
// TODO(zcfgs-pod): We want to validate for secondary tenants as
// well, but they don't (yet) have access to node localites.
```
</blockquote></details>

Done.
___
*[pkg/sql/show_zone_config.go, line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSrRQKA-mVg5nqs-r3-124:-MeQhR463d0IgWc9T3b0:b60rai3) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/sql/show_zone_config.go#L124)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

If we remove the `if !p.ExecCfg().Codec.ForSystemTenant() {` check in `ShowZoneConfig`, we should "just" get introspection.
</blockquote></details>

Yup, tacked on in a new commit. 
___
*[pkg/sql/zone_config.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSrRQKA-mVg5nqt-r3-58:-MeQPqIN6pNd9lqoOvo9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/sql/zone_config.go#L58)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Swap the order of arguments, codec typically precedes what's being decoded through it.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
Haven't taken a full look, but for seeding the default zone config, try this instead:
```diff
diff --git i/pkg/sql/catalog/bootstrap/metadata.go w/pkg/sql/catalog/bootstrap/metadata.go
index 08ddcd13cb..bfabb4460c 100644
--- i/pkg/sql/catalog/bootstrap/metadata.go
+++ w/pkg/sql/catalog/bootstrap/metadata.go
@@ -355,48 +355,50 @@ func addZoneConfigKVsToSchema(
 	defaultZoneConfig *zonepb.ZoneConfig,
 	defaultSystemZoneConfig *zonepb.ZoneConfig,
 ) {
-	// If this isn't the system tenant, don't add any zone configuration keys.
-	// Only the system tenant has a zone table.
-	if !target.codec.ForSystemTenant() {
-		return
-	}
-
-	// Adding a new system table? It should be added here to the metadata schema,
-	// and also created as a migration for older cluster. The includedInBootstrap
-	// field should be set on the migration.
+	// Adding a new system table? It should be added here to the metadata
+	// schema, and also created as a migration for older cluster. The
+	// includedInBootstrap field should be set on the migration.
 
+	// All tenants have their own RANGE DEFAULT zone config.
 	target.otherKV = append(target.otherKV,
 		createZoneConfigKV(keys.RootNamespaceID, target.codec, defaultZoneConfig))
 
-	systemZoneConf := defaultSystemZoneConfig
-	metaRangeZoneConf := protoutil.Clone(defaultSystemZoneConfig).(*zonepb.ZoneConfig)
-	livenessZoneConf := protoutil.Clone(defaultSystemZoneConfig).(*zonepb.ZoneConfig)
-
-	// .meta zone config entry with a shorter GC time.
-	metaRangeZoneConf.GC.TTLSeconds = 60 * 60 // 1h
+	// All tenants have their own DATABASE SYSTEM zone config.
 	target.otherKV = append(target.otherKV,
-		createZoneConfigKV(keys.MetaRangesID, target.codec, metaRangeZoneConf))
+		createZoneConfigKV(keys.SystemDatabaseID, target.codec, defaultZoneConfig))
 
-	// Some reporting tables have shorter GC times.
-	replicationConstraintStatsZoneConf := &zonepb.ZoneConfig{
-		GC: &zonepb.GCPolicy{TTLSeconds: int32(systemschema.ReplicationConstraintStatsTableTTL.Seconds())},
-	}
-	replicationStatsZoneConf := &zonepb.ZoneConfig{
-		GC: &zonepb.GCPolicy{TTLSeconds: int32(systemschema.ReplicationStatsTableTTL.Seconds())},
-	}
+	if target.codec.ForSystemTenant() {
+		// Only the system tenant has zone configs declared over RANGE
+		// {META,LIVENESS,SYSTEM} and over TABLE
+		// system.public.replication{,_constraint}_stats.
 
-	// Liveness zone config entry with a shorter GC time.
-	livenessZoneConf.GC.TTLSeconds = 10 * 60 // 10m
-	target.otherKV = append(target.otherKV,
-		createZoneConfigKV(keys.LivenessRangesID, target.codec, livenessZoneConf))
-	target.otherKV = append(target.otherKV,
-		createZoneConfigKV(keys.SystemRangesID, target.codec, systemZoneConf))
-	target.otherKV = append(target.otherKV,
-		createZoneConfigKV(keys.SystemDatabaseID, target.codec, systemZoneConf))
-	target.otherKV = append(target.otherKV,
-		createZoneConfigKV(keys.ReplicationConstraintStatsTableID, target.codec, replicationConstraintStatsZoneConf))
-	target.otherKV = append(target.otherKV,
-		createZoneConfigKV(keys.ReplicationStatsTableID, target.codec, replicationStatsZoneConf))
+		// .meta zone config entry with a shorter GC time.
+		metaRangeZoneConf := protoutil.Clone(defaultSystemZoneConfig).(*zonepb.ZoneConfig)
+		metaRangeZoneConf.GC.TTLSeconds = 60 * 60 // 1h
+		target.otherKV = append(target.otherKV,
+			createZoneConfigKV(keys.MetaRangesID, target.codec, metaRangeZoneConf))
+
+		// .liveness zone config entry with a shorter GC time.
+		livenessZoneConf := protoutil.Clone(defaultSystemZoneConfig).(*zonepb.ZoneConfig)
+		livenessZoneConf.GC.TTLSeconds = 10 * 60 // 10m
+		target.otherKV = append(target.otherKV,
+			createZoneConfigKV(keys.LivenessRangesID, target.codec, livenessZoneConf))
+
+		target.otherKV = append(target.otherKV,
+			createZoneConfigKV(keys.SystemRangesID, target.codec, defaultSystemZoneConfig))
+
+		// Reporting tables have shorter GC times.
+		replicationConstraintStatsZoneConf := &zonepb.ZoneConfig{
+			GC: &zonepb.GCPolicy{TTLSeconds: int32(systemschema.ReplicationConstraintStatsTableTTL.Seconds())},
+		}
+		replicationStatsZoneConf := &zonepb.ZoneConfig{
+			GC: &zonepb.GCPolicy{TTLSeconds: int32(systemschema.ReplicationStatsTableTTL.Seconds())},
+		}
+		target.otherKV = append(target.otherKV,
+			createZoneConfigKV(keys.ReplicationConstraintStatsTableID, target.codec, replicationConstraintStatsZoneConf))
+		target.otherKV = append(target.otherKV,
+			createZoneConfigKV(keys.ReplicationStatsTableID, target.codec, replicationStatsZoneConf))
+	}
 }
 
 // addSystemDatabaseToSchema populates the supplied MetadataSchema with the
diff --git i/pkg/sql/tenant.go w/pkg/sql/tenant.go
index e5414b4bd3..ae234a28b4 100644
--- i/pkg/sql/tenant.go
+++ w/pkg/sql/tenant.go
@@ -156,10 +156,11 @@ func (p *planner) CreateTenant(ctx context.Context, tenID uint64) error {
 
 	codec := keys.MakeSQLCodec(roachpb.MakeTenantID(tenID))
 	// Initialize the tenant's keyspace.
+	// XXX: Do we want these to be overridable?
 	schema := bootstrap.MakeMetadataSchema(
 		codec,
-		nil, /* defaultZoneConfig */
-		nil, /* defaultZoneConfig */
+		zonepb.DefaultZoneConfigRef(),
+		zonepb.DefaultSystemZoneConfigRef(),
 	)
 	kvs, splits := schema.GetInitialValues()
 
@@ -179,15 +180,6 @@ func (p *planner) CreateTenant(ctx context.Context, tenID uint64) error {
 		kvs = append(kvs, tenantSettingKV)
 	}
 
-	{
-		// Seed the tenant with a default zone configuration.
-		tenantZoneConfigKVs, err := generateTenantDefaultZoneConfigKVs(codec)
-		if err != nil {
-			return err
-		}
-		kvs = append(kvs, tenantZoneConfigKVs...)
-	}
-
 	b := p.Txn().NewBatch()
 	for _, kv := range kvs {
 		b.CPut(kv.Key, &kv.Value, nil)
@@ -260,44 +252,6 @@ func generateTenantClusterSettingKV(
 	}, nil
 }
 
-func generateTenantDefaultZoneConfigKVs(codec keys.SQLCodec) ([]roachpb.KeyValue, error) {
-	defaultZoneConfig := zonepb.DefaultZoneConfig()
-	defaultZoneConfigProto, err := protoutil.Marshal(&defaultZoneConfig)
-	if err != nil {
-		return nil, err
-	}
-	kvs, err := rowenc.EncodePrimaryIndex(
-		codec,
-		systemschema.ZonesTable,
-		systemschema.ZonesTable.GetPrimaryIndex(),
-		catalog.ColumnIDToOrdinalMap(systemschema.ZonesTable.PublicColumns()),
-		[]tree.Datum{
-			tree.NewDInt(keys.RootNamespaceID),                  // id
-			tree.NewDBytes(tree.DBytes(defaultZoneConfigProto)), // config
-		},
-		false, /* includeEmpty */
-	)
-	if err != nil {
-		return nil, errors.NewAssertionErrorWithWrappedErrf(
-			err, "failed to encode default zone config",
-		)
-	}
-	if len(kvs) != 2 {
-		return nil, errors.AssertionFailedf(
-			"failed to encode default zone config: expected 2 column families, got %d %v", len(kvs), kvs)
-	}
-	return []roachpb.KeyValue{
-		{
-			Key:   kvs[0].Key,
-			Value: kvs[0].Value,
-		},
-		{
-			Key:   kvs[1].Key,
-			Value: kvs[1].Value,
-		},
-	}, nil
-}
-
 // ActivateTenant marks a tenant active.
 func ActivateTenant(ctx context.Context, execCfg *ExecutorConfig, txn *kv.Txn, tenID uint64) error {
 	const op = "activate"
```

Like discussed offline, we'll also want a corresponding long running migration that adds these default rows for existing tenants.

irfansharif reviewed (approved):
1. Collapse everything into a single commit, it'll make it much easier when porting over to master. 
2. Let's use TODO(zcfgs-pod) for things we need to address as part of the prototype, it'll help distinguish between other TODOs elsewhere in the code with our names attached. There are a few instances in this PR.
3. Let's simplify the default version bootstrap code per the previous top-level comment above, or something similar to it.

> Reviewed 2 of 8 files at r1, 2 of 17 files at r2, 5 of 7 files at r3, 14 of 16 files at r4, 2 of 12 files at r7, 2 of 2 files at r8, 9 of 9 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-:-Meaf1phBZu5NDZath15:blud7ga)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[pkg/ccl/logictestccl/testdata/logic_test/tenant, line 48 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-MeaaXhN23niB7Czjb7P:-MeaaXhN23niB7Czjb7Q:b-g4owmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/062b5872209ac9ad2e57a3d6d8db6d531634af39/pkg/ccl/logictestccl/testdata/logic_test/tenant#L48)):*
> ```
> SELECT * FROM crdb_internal.cluster_contention_events
> 
> # Selecting from crdb_internal.zones is allowed.
> ```

Looking at the other tests in this file + us having enabled the tenant config for the zone_config logic tests, are these the right kind of tests to be adding here? If we're validating in this test file whether this is an allowable operation, I think just the following is sufficient:

```
statement ok
SELECT * FROM crdb_internal.zones
```
___
*[pkg/ccl/logictestccl/testdata/logic_test/tenant, line 50 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Meabz1oAIQY-VgrN2ND:-Meabz1oAIQY-VgrN2NE:bv1rhfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/062b5872209ac9ad2e57a3d6d8db6d531634af39/pkg/ccl/logictestccl/testdata/logic_test/tenant#L50)):*
> ```
> # Selecting from crdb_internal.zones is allowed.
> query IITTTTTTTTTTTT
> SELECT * FROM crdb_internal.zones
> ```

Delete this test from this file, it's better placed elsewhere (per comment above).
___
*[pkg/config/keys.go, line 36 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSq-McpWkamiJMV-r3-36:-Meac_td9lnhD5N2cI5Z:b-kg5aat) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/config/keys.go#L36)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

> Better yet, can we get rid of the SystemTenantObjectID type entirely?

See my comment below about this. I don't think we can as part of this PR, but we should work towards this eventually. 

> This function should also thread in a codec. Does it have to return this SystemTenantObjectID type?

Considering this thing is only called through from codepaths in kvserver, I was thinking generalizing that stuff would serve as a natural place to do this. Do you think that's reasonable? 
</blockquote></details>

For codepaths in kvserver that end up with SystemTenantObjectID, I figured we could just explicitly thread in the SystemSQLCodec instead -- we're doing that elsewhere. I'm fine with this PR as is too, and looks like we both agree on the end-state.
___
*[pkg/config/system.go, line 538 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSq-McpWkamiJMW-r3-538:-Mead1isBxoCw80kmcdu:b-h104xs) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/config/system.go#L538)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I was hoping to isolate codec threading to just the SQL code required to set/introspect zone configs. Considering `DecodeKeyIntZoneIDAndSuffix` is primarily called from the replication report code, I think it makes sense to thread in the codec once we're tackling that. Wdyt? 
</blockquote></details>

See comment above, I think it's more than fine to thread in a SystemSQLCode from the replication report code, if it gives us the short term win of removing the SystemTenantObjectID concept. Is it that big a change?
___
*[pkg/sql/tenant.go, line 184 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Meae0jI8gj8CD4ZOUC8:-Meae0jI8gj8CD4ZOUC9:b-5uz2l9) ([raw file](https://github.com/cockroachdb/cockroach/blob/062b5872209ac9ad2e57a3d6d8db6d531634af39/pkg/sql/tenant.go#L184)):*
> ```Go
> 	{
> 		// Seed the tenant with a default zone configuration.
> 		tenantZoneConfigKVs, err := generateTenantDefaultZoneConfigKVs(codec)
> ```

See top-level comment for the bootstrap simplification we should do instead. Separately: as discussed offline, add a TODO(zcfgs-pod) comment outlining why this kind of thing is much better suited to a single startup migration, instead of adding it in two places (here for new tenants, and a tenant migration for existing tenants).
___
*[pkg/sql/zone_config.go, line 197 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSrRQKA-mVg5nqt-r3-197:-MeaeP36EGYuTdY3cRsw:b-x5xxtm) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/sql/zone_config.go#L197)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I should've done a better job adding words to that TODO haha. I'll try to add more context on what the state was before my change, why I chose this as the stopping point, and where I see this going once we have this whole thing wired up. 

At a high level, all zone config code is organized in a way that it goes through this `getZoneConfig` function. When KV accesses it using this hook, `getKey` reads the system config span. Other SQL code does transactional reads. Previously, SQL code was gated to only let the system tenant through. As such, we could cast descriptor IDs to `SystemTenantObjectIDs`. This is no longer true after my change, so I've generalized all code in SQL to require codecs and work for arbitrary object IDs. 

This function here, that KV calls, is one of the few places in SQL that still uses `SystemTenantObjectID`. Once we've generalized things in kvserver and it no longer calls into SQL (and instead reference the span configs table), we'll get rid of this whole `zoneConfigHook` entirely. Until then, I have this assertion baked in here. 

Once we've generalized everything in kvserver/reporter I think we can get rid of `SystemTenantObjectID` in its entirety. That's why removing (most) occurrences in the sql package felt like a good stopping point. 

Separately, one of the things that makes this zone config code super hairy is this `getKey` business and everything funnelling to `getZoneConfig`. I'd really like to simplify all this once we've gotten rid of this `zoneConfigHook`, but that'll be mostly cleanup.  

</blockquote></details>

Ack. LGTM as an incremental step towards getting rid of SystemTenantObjectID entirely.
___
*[pkg/sql/logictest/testdata/logic_test/zone_config, line 1 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Meaebat1VpX1cEtXAv0:-Meaebat1VpX1cEtXAv1:b-83sjo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ffc51d4a844003af6938be26d9c8978fb818d7a/pkg/sql/logictest/testdata/logic_test/zone_config#L1)):*
> ```
> # LogicTest: !3node-tenant(49854)
> ```

Nice.
___
*[pkg/sql/logictest/testdata/logic_test/zone_config_system_tenant, line 7 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Meaedi96oRmK3b9LsF3:-Meaedi96oRmK3b9LsF4:b-cyvjaf) ([raw file](https://github.com/cockroachdb/cockroach/blob/062b5872209ac9ad2e57a3d6d8db6d531634af39/pkg/sql/logictest/testdata/logic_test/zone_config_system_tenant#L7)):*
> ```
> CREATE TABLE a (id INT PRIMARY KEY)
> 
> # TODO(arul): This should work once secondary tenants can validate underlying
> ```

TODO(zcfgs-pod). I assume we'll get this for free rebasing on top of #67098.


<!-- Sent from Reviewable.io -->

arulajmani committed:
sql: enable secondary tenants to write zone configurations

This patch adds the `system.zones` table for secondary tenants and
allows them to modify and introspect zone configurations. As a result,
 we can now run the zone_config logictest for tenants as well.

Release note: None

arulajmani reviewed (commented):
Done -- I'll merge this once CI is green 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-:-MefRHBX4J0aCz2sCeJr:b95pk2h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @irfansharif)

___
*[pkg/config/system.go, line 538 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Me1GDSq-McpWkamiJMW-r3-538:-Meev59u5pv1BWS4XrTf:b8h5nu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/338e6a71380e4cf45beed4dc0ca72ddbb2efc4d9/pkg/config/system.go#L538)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

See comment above, I think it's more than fine to thread in a SystemSQLCode from the replication report code, if it gives us the short term win of removing the SystemTenantObjectID concept. Is it that big a change?
</blockquote></details>

We're not really getting rid of the SystemTenantObjectID entirely though, are we? We're just making it implicit -- the code would still only work for the system tenant. Keeping it around makes it easy to find all the places that still need to be generalized. Also, it's pretty cheap to keep around -- you just need to a cast and pass in the system codec. 
___
*[pkg/sql/tenant.go, line 184 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Meae0jI8gj8CD4ZOUC8:-MeedhGsD05TCAxRnLuh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/062b5872209ac9ad2e57a3d6d8db6d531634af39/pkg/sql/tenant.go#L184)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

See top-level comment for the bootstrap simplification we should do instead. Separately: as discussed offline, add a TODO(zcfgs-pod) comment outlining why this kind of thing is much better suited to a single startup migration, instead of adding it in two places (here for new tenants, and a tenant migration for existing tenants).
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/zone_config_system_tenant, line 7 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Meaedi96oRmK3b9LsF3:-MeecYVeDFbOU60YBb12:b2nhtan) ([raw file](https://github.com/cockroachdb/cockroach/blob/062b5872209ac9ad2e57a3d6d8db6d531634af39/pkg/sql/logictest/testdata/logic_test/zone_config_system_tenant#L7)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

TODO(zcfgs-pod). I assume we'll get this for free rebasing on top of #67098.
</blockquote></details>

Yup
___
*[pkg/ccl/logictestccl/testdata/logic_test/tenant, line 48 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-MeaaXhN23niB7Czjb7P:-MeepXos1EbnOY0cbgJi:bxgv83l) ([raw file](https://github.com/cockroachdb/cockroach/blob/062b5872209ac9ad2e57a3d6d8db6d531634af39/pkg/ccl/logictestccl/testdata/logic_test/tenant#L48)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Looking at the other tests in this file + us having enabled the tenant config for the zone_config logic tests, are these the right kind of tests to be adding here? If we're validating in this test file whether this is an allowable operation, I think just the following is sufficient:

```
statement ok
SELECT * FROM crdb_internal.zones
```
</blockquote></details>

To be fair this was always an allowable operation, even when it was in `tenant_unsupported` -- it just returned no results. I had it here because it seems like this file is a dumping ground for things that move out of `tenant_unsupported`, but I guess I'll slightly change this test and put it in a new file `zone_config_secondary_tenants`. 

It's worth mentioning that turning on `zone_config` logictests doesn't give us coverage against this `crdb_internal.zones` table -- the only logic test that seems exercise this is in `crdb_internal` and that thing is only run in the local configuration. It's also not possible to run that entire file in 3node-tenant, because the results of this select will change depending on whether we're dealing with a secondary tenant or the system tenant. 
___
*[pkg/ccl/logictestccl/testdata/logic_test/tenant, line 50 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/67323#-Meabz1oAIQY-VgrN2ND:-MeeuywYDKHCkDlWBMx_:bbx87r0) ([raw file](https://github.com/cockroachdb/cockroach/blob/062b5872209ac9ad2e57a3d6d8db6d531634af39/pkg/ccl/logictestccl/testdata/logic_test/tenant#L50)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Delete this test from this file, it's better placed elsewhere (per comment above).
</blockquote></details>

See comment above. 


<!-- Sent from Reviewable.io -->

arulajmani commented:
As discussed offline, there's a couple things we still want to change here. Namely, we want to switch from a long running migration to a startup migration here -- that seems more suited to this on second thought. This is also missing a default zone config for the system database. I'm merging this for now but I'll come back with a separate PR with these changes considering it's a prototype branch.
